internal/ingest/linksystem.go

```diff
@@ -204,35 +213,146 @@ func (ing *Ingester) ingestAd(publisherID peer.ID, adCid cid.Cid, ad schema.Adve
 
 	startTime := time.Now()
 
-	// Traverse entries based on the entries selector that limits recursion depth.
+	// The ad.Entries link can point to either a chain of EntryChunks or a HAMT.
+	// Sync the very first entry so that we can check which type it is.
+	// Note, this means the maximum depth of entries traversal will be 1 plus the configured max depth.
+	// TODO: See if it is worth detecting and reducing depth the depth in entries selectors by one.
+	syncedFirstEntryCid, err := ing.sub.Sync(ctx, publisherID, entriesCid, Selectors.One, nil)
+	if err != nil {
+		return adIngestError{adIngestSyncEntriesErr, fmt.Errorf("failed to sync first entry while checking entries type: %w", err)}
+	}
+
+	node, err := ing.loadNode(syncedFirstEntryCid, basicnode.Prototype.Any)
+	if err != nil {
+		return adIngestError{adIngestIndexerErr, fmt.Errorf("failed to load first entry after sync: %w", err)}
+	}
+
 	var errsIngestingEntryChunks []error
-	_, err = ing.sub.Sync(ctx, publisherID, entriesCid, ing.entriesSel, nil, legs.ScopedBlockHook(func(p peer.ID, c cid.Cid, actions legs.SegmentSyncActions) {
-		// Load CID as entry chunk since the selector should only select entry chunk nodes.
-		chunk, err := ing.loadEntryChunk(c)
+	if isHAMT(node) {
+		log = log.With("entriesKind", "hamt")
+		// Keep track of all CIDs in the HAMT to remove them later when the processing is done.
+		// This is equivalent behaviour to ingestEntryChunk which removes an entry chunk right afrer
+		// it is processed.
+		hamtCids := []cid.Cid{syncedFirstEntryCid}
+		gatherCids := func(_ peer.ID, c cid.Cid, _ legs.SegmentSyncActions) {
+			hamtCids = append(hamtCids, c)
+		}
+		defer func() {
```

Comment:
i don't see where this method would have previously remove entry chunks before return,
and i don't see where it does it in the non-hamt case. Why is it different here?

Reply:
[Here](https://github.com/filecoin-project/storetheindex/blob/main/internal/ingest/linksystem.go#L259-L268) it is.